### DISCUSSION TITLE
Feature suggestion: Helper function to find errors nested in transactions

### DISCUSSION THREAD

**Discussion by Tbaut** [polkadot-api/polkadot-api#700] [Ideas]
Hello there,

Here's a non urgent suggestion, to add some function to help Dapp devs to display issues deeply nested into Txs that didn't succeed totally. 

There are cases, such as multisig, batch, sudo.. or probably any extrinsic that has a `call` param, where the extrinsic may appear to be a success, but the underlying calls failed. If I do a batch, and some call fails in there, as a user, I want to know.

For the Dapp dev to show these errors, I need to know all the calls I'm doing in detail. Below is an example of a batch with a `System.remark` (works) and a `System.set_code` (Bad Origin). To show to my users, that the code failed, I need to:
- check the events
- look for type `Utility` and `value.type` is `BatchInterrupted` and `value.value.error.type` is `BadOrigin`
- display the `BadOrigin`

Now for multisig, the error is in a different type, same for sudo, for proxy etc..

This is a lot of clever code that probably needs to be repeated on many Dapps, and it would be very valuable to have 1 registry for this.

Here's a call example, with the batch I described, for westend. Please just put in your mnemonic.

```ts
import { sr25519CreateDerive } from '@polkadot-labs/hdkd';
import {
  entropyToMiniSecret,
  mnemonicToEntropy,
} from '@polkadot-labs/hdkd-helpers';
import { getPolkadotSigner } from 'polkadot-api/signer';
import { wnd } from '@polkadot-api/descriptors';
import { Binary, createClient } from 'polkadot-api';
import { getWsProvider } from 'polkadot-api/ws-provider/node';

const main = async () => {
  const aliceKeyPair = sr25519CreateDerive(
    entropyToMiniSecret(mnemonicToEntropy('YOUR_MNEMONIC_HERE')),
  )('');

  const aliceSigner = getPolkadotSigner(
    aliceKeyPair.publicKey,
    'Sr25519',
    aliceKeyPair.sign,
  );

  // CLIENT AND API
  const client = createClient(
    getWsProvider('wss://rpc.dotters.network/westend'),
  );
  const api = client.getTypedApi(wnd);

  // create the call from some call-data
  // here a batch with a remark, and a system.setCode that we're not allowed to do
  const tx = await api.txFromCallData(
    Binary.fromHex('0x10000800002c73686f756c6420776f726b000208beef'),
  );
  // create the signed extrinsic
  const signedTx = await tx.sign(aliceSigner);
  // see what the result of this extrinsic would be against the current best-block
  const dryRunResult = await api.apis.BlockBuilder.apply_extrinsic(
    Binary.fromOpaqueHex(signedTx),
    { at: 'best' },
  );
  // `dryRunResult` is a strongly typed object, so if `success` is false, then
  // the value will have a strongly typed enum with the reason why it didn't succeed
  console.log(dryRunResult);
  // { success: true, value: { success: true, value: undefined } }

  // The the dry run is successful, but there are errors in the events
  // although the dispatchError is undefined
  if (dryRunResult.success) {
    const submitResult = await client
      .submitAndWatch(signedTx)
      .subscribe((event) => {
        if (event.type === 'txBestBlocksState' && event.found === true) {
          console.log('dispatchError', event.dispatchError);
          event.events.forEach((e) =>
            console.log(
              JSON.stringify(e, (_, v) =>
                typeof v === 'bigint' ? v.toString() : v,
              ),
            ),
          );
        }

        //  dispatchError undefined
        // {"type":"Balances","value":{"type":"Withdraw","value":{"who":"5HbVkMa1pk2aDBv7CGGyN3AQYjy9W3wwGRPd8kSXF1eAVpdX","amount":"127369831879"}}}
        // {"type":"Utility","value":{"type":"ItemCompleted"}}
        // {"type":"Utility","value":{"type":"BatchInterrupted","value":{"index":1,"error":{"type":"BadOrigin"}}}}
        // {"type":"Balances","value":{"type":"Deposit","value":{"who":"5HbVkMa1pk2aDBv7CGGyN3AQYjy9W3wwGRPd8kSXF1eAVpdX","amount":"0"}}}
        // {"type":"Balances","value":{"type":"Deposit","value":{"who":"5G1ojzh47Yt8KoYhuAjXpHcazvsoCXe3G8LZchKDvumozJJJ","amount":"127369831879"}}}
        // {"type":"TransactionPayment","value":{"type":"TransactionFeePaid","value":{"who":"5HbVkMa1pk2aDBv7CGGyN3AQYjy9W3wwGRPd8kSXF1eAVpdX","actual_fee":"127369831879","tip":"0"}}}
        // {"type":"System","value":{"type":"ExtrinsicSuccess","value":{"dispatch_info":{"weight":{"ref_time":"129165303376","proof_size":"1485"},"class":{"type":"Normal"},"pays_fee":{"type":"Yes"}}}}}
      });
  } else {
    console.log('dryrun failed');
  }
};

main();

```

**Comment by carlosala** [comment]
Hi! [A `utility.batch` is always successful, it can't fail.](https://github.com/paritytech/polkadot-sdk/blob/986e7ae4f29f804ee4dc89aaf52984d6eda5bd0b/substrate/frame/utility/src/lib.rs#L166-L183) If you want you call to fail and avoid executing any transaction if not every was successful, [you have to use `batch_all`](https://github.com/paritytech/polkadot-sdk/blob/986e7ae4f29f804ee4dc89aaf52984d6eda5bd0b/substrate/frame/utility/src/lib.rs#L293-L305).

I'm afraid your proposition is outside of this library's boundaries. We specifically don't implement any pallet or call business logic, and that's the responsibility of the dApp developer. It could be a great fit for our pallet-specific SDKs, that are on the pipeline, but not for the core `polkadot-api`.
Thanks for the idea!

**Reply by josepot** [reply]
I don't necessarily agree with this. I think that there are a lot of useful utils that we should provide that are not tightly coupled to these SDKs. Utils for things like: formatting currencies, parsing currencies, stringifying objects that contain bigints, etc, etc. I am very open to the idea of having a helper function that helps at segregating the events of the "nested" transactions. It's tricky, but perhaps we can find a way.

It is true that the SDKs would offer more value, but just because we offer a higher layer that already solves this issue nicely, that doesn't mean that we shouldn't expose these kinds of utils.

I need to put some more thought into this, but I just wanted to make it clear that the SDKs are not at odds with these kinds of helpers

**Comment by carlosala** [comment]
Converting it into a discussion since it is not an issue of the core `polkadot-api`.

**Comment by Tbaut** [comment]
At the end of the day, I thought about opening an issue, because when I'm switching from pjs to papi these days, I've keep in my head the question: "what could the papi team do, that would help me, and probably many other devs, make the process of building Dapps so much less painful". 

I thought about it when I saw the horrendous pjs code I had, and that I have to migrate now. I know I'll spend no fun hours figuring out all the potential error types etc, and unfortunately unless other devs know I did it, and take my code, they'll probably do it again. So I thought I'd at least let you know, and although this is absolutely not a priority, it's good that you heard it.

BTW, 2 side questions as I'm going through this

- 1 how would you go about trying to find out about the potential errors, and how to get to the type, e.g, I need some mapping , basically like[ this one](https://github.com/polkadot-js/apps/blob/97aeb41a593165f9a9b218c1ceecb52172473de3/packages/react-components/src/Status/checks.ts)

For batch I know I can do:
```ts
        if (event.type === 'Utility' && event.value.type === 'BatchInterrupted') {
          errorInfo = event.value.value.error.type
        }
```
How do I find out the Multisig ones for instance without reproducing them, [it's weirdly nested from what I can see](https://github.com/polkadot-js/apps/blob/97aeb41a593165f9a9b218c1ceecb52172473de3/packages/react-components/src/Status/checks.ts#L36-L39). 
edit: going through the common-types proves being quite difficult and error prone.


- 2 is there a way to have actually nice error messages, such as the ones displayed in this table https://wiki.polkadot.network/docs/maintain-errors#error-table I guess this isn't in the metadata, it's some off-chain registry I imagine.
